### PR TITLE
Coveralls patch

### PR DESCRIPTION
**Pull request by ranok**
This patch (once merged and someone with repo access to the monero-project's bitmonero has created a free coveralls.io page for the repo) will provide the badge on the readme that shows the test coverage and will automatically recompute per push to measure improvements/degradation. The biggest challenge is the time limit enforced by Travis-CI (50min) that prevents core_tests from being finished, so if it's being run in Travis, those tests are skipped. I have also tidied up the travis YAML into using a build matrix, one build for testing, one for g++ and one for clang.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -392,9 +392,9 @@ else()
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DGTEST_HAS_TR1_TUPLE=0")
   endif()
   if(CMAKE_C_COMPILER_ID STREQUAL "GNU" AND NOT (CMAKE_C_COMPILER_VERSION VERSION_LESS 4.8))
-    set(DEBUG_FLAGS "-g3 -Og")
+    set(DEBUG_FLAGS "-g3 -Og -Wall -fprofile-arcs -ftest-coverage --coverage")
   else()
-    set(DEBUG_FLAGS "-g3 -O0")
+    set(DEBUG_FLAGS "-g3 -O0 -Wall -fprofile-arcs -ftest-coverage --coverage")
```
Comment (moneromooo-monero) [inline CMakeLists.txt]: That looks like there should be a separate "profile" target, in addition to release and debug

File: .travis.yml [hunk 2]
```diff
@@ -1,35 +1,110 @@
...
+    before_install:
+      - sudo add-apt-repository -y ppa:kojoley/boost
+      - sudo apt-get -q update
+    install:
+      - sudo apt-get -y install libboost-{chrono,program-options,date-time,thread,system,filesystem,regex,serialization}1.58{-dev,.0}
+    script:
+      - make -j2 && HAVE_DOT=YES doxygen Doxyfile
+
+  #
+  # Monero release-all (clang)
```
Comment (moneromooo-monero) [inline .travis.yml]: Do these comments get parsed ? I see no other distinction for build types... Just a question, no chnage request

**Comment by moneromooo-monero** [comment]
just a comment about maybe prefering a separate profile (or coverage) target. I'm not sure if coverage would kill performance even more than debug. Otherwise ok.

File: .travis.yml (hunk 2 above)
Comment (radfish) [inline .travis.yml]: The '-' does it, doesn't it? List element of the 'matrix' list. (Just guessing).

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -54,7 +54,10 @@ else ()
   endif()
 endif ()
 
-add_subdirectory(core_tests)
+# Skip the core_tests if we are running in Travis-CI because they will take too long
+if (NOT DEFINED ENV{TRAVIS})
```
Comment (radfish) [inline tests/CMakeLists.txt]: Is this special-case necessary, given the later special-case around add_custom_target? If not, I would remove it, to minimize the special-cases and also to leave open the option to manually invoke the target to build this test (the extra time for cmake to parse that subdir is tiny).

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -63,7 +66,9 @@ add_subdirectory(unit_tests)
 add_subdirectory(difficulty)
 add_subdirectory(hash)
 add_subdirectory(net_load_tests)
-add_subdirectory(libwallet_api_tests)
+
+# Disabled until issue #895 is resolved
+#add_subdirectory(libwallet_api_tests)
```
Comment (radfish) [inline tests/CMakeLists.txt]: Why disable tests here? Isn't this exactly the wonderful benefit of all your work? The badge will show that the test is failing and will put a little extra pressure towards fixing the root cause.

File: CMakeLists.txt (hunk 1 above)
Comment (radfish) [inline CMakeLists.txt]: There is logic elsewhere that appends -Wall to *FLAGS, see `WARNINGS` var. Could we avoid duplication? If somehow -Wall doesn't end up where you want it, let's investigate why.

Agreed on either a separating profiling. How about add a cmake option (with `option()`) for this and set it to false by default and set it to true in a new 'profile' target in top-level makefile?

**Comment by radfish** [comment]
A question: is 50min per element in the build matrix or total? I assume it's per element because 3 builds take more than 50mins. If it's per element, how hard would it be to add two gcc versions (6.1 and 4.8)? Previously tests broke on gcc 6.1 but not on gcc 4.8 due to aliasing.

Question: is it possible/desirable to add coverage stats to all builds in the matrix? So, instead of {gcc-v4,gcc-v6,clang,coverage} have {gcc-v4-with-coverage,gcc-v6-with-coverage,clang-with-coverage}. It's not unreasonable to expect tests to fail for some subset of compilers/versions only.

**Comment by ranok** [comment]
@radfish: The 50min is per "job", which is each entry. I can certainly other compilers, and move towards that model of a coverage build for each compiler

File: tests/CMakeLists.txt (hunk 4 above)
Comment (ranok) [inline tests/CMakeLists.txt]: All the documentation I've seen put the coveralls reporting in the after_success step which is not reached if any tests fail. I will experiment and see if there are any issues reporting coverage for a failing test suite.

File: .travis.yml (hunk 2 above)
Comment (ranok) [inline .travis.yml]: @moneromooo-monero You are right, the comments in the travis config are just to clarify what each of the build tasks are doing

**Comment by radfish** [comment]
Ok. On a second thought, coverage per compiler is not valuable. We can get information on pass/fail for each test from the regular builds, instead of from the coverage build. Sorry to create needless confusion.

File: .travis.yml (hunk 2 above)
Comment (moneromooo-monero) [inline .travis.yml]: Nevermind, I was expecting some kind of header for each, but they're just unnamed it seems. Ignore what I said :)

**Pull request merged**